idx/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,441 @@
+package elasticsearch
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mattbaird/elastigo/lib"
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/metrictank/idx/memory"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxEsOk             met.Count
+	idxEsFail           met.Count
+	idxEsAddDuration    met.Timer
+	idxEsDeleteDuration met.Timer
+
+	Enabled        bool
+	Index          string
+	Hosts          string
+	User           string
+	Pass           string
+	RetryInterval  string
+	MaxConns       int
+	MaxBufferDocs  int
+	BufferDelayMax string
+)
+
+func ConfigSetup() {
+	esIdx := flag.NewFlagSet("elasticsearch-idx", flag.ExitOnError)
+	esIdx.BoolVar(&Enabled, "enabled", false, "")
+	esIdx.StringVar(&Index, "index", "metric", "Elasticsearch index name for storing metric index.")
+	esIdx.StringVar(&Hosts, "hosts", "localhost:9200", "comma separated list of elasticsearch address in host:port form")
+	esIdx.StringVar(&User, "user", "", "HTTP basic Auth username")
+	esIdx.StringVar(&Pass, "pass", "", "HTTP basic Auth password")
+	esIdx.StringVar(&RetryInterval, "retry-interval", "1h", "Interval to retry indexing Definitions in ES after failures")
+	esIdx.IntVar(&MaxConns, "max-conns", 20, "Max number of http conns in flight to ES servers at one time")
+	esIdx.IntVar(&MaxBufferDocs, "max-buffer-docs", 1000, "Max number of Docs to hold in buffer before forcing flush to ES")
+	esIdx.StringVar(&BufferDelayMax, "buffer-delay-max", "10s", "Max time to wait before flusing forcing flush to ES")
+	globalconf.Register("elasticsearch-idx", esIdx)
+}
+
+type RetryBuffer struct {
+	Index       *EsIdx
+	Defs        []schema.MetricDefinition
+	LastAttempt time.Time
+	done        chan struct{}
+	sync.Mutex
+}
+
+func NewRetryBuffer(index *EsIdx, interval time.Duration) *RetryBuffer {
+	r := &RetryBuffer{
+		Index:       index,
+		Defs:        make([]schema.MetricDefinition, 0),
+		LastAttempt: time.Now(),
+		done:        make(chan struct{}),
+	}
+	go r.run(interval)
+	return r
+}
+
+func (r *RetryBuffer) Stop() {
+	close(r.done)
+}
+
+func (r *RetryBuffer) Items() []schema.MetricDefinition {
```

Comment:
should not be a public function. and since it's only being used to verify the number contained in the buffer, why don't we just turn it into a function `num()` or something?